client/db/benches/state_access.rs

```diff
@@ -104,12 +136,28 @@ fn insert_blocks(db: &Backend<Block>, storage: Vec<(Vec<u8>, Vec<u8>)>) -> H256
 	parent_hash
 }
 
+#[derive(Copy, Clone)]
 enum BenchmarkConfig {
 	NoCache,
 	TrieNodeCache,
 }
 
+#[derive(Copy, Clone)]
+enum AccessType {
+	Storage,
+	StorageHash,
+}
+
 fn create_backend(config: BenchmarkConfig, temp_dir: &TempDir) -> Backend<Block> {
+	// To prevent a "Database file is in use" error, since backend's
+	// deinitialization is not synchronous.
+	//
+	// Technically we could recreate everything for each benchmark,
+	// but that takes forever as we'd have to repopulate the database
+	// on disk from scratch every time. It's a lot faster if we just
+	// wait for it to close and reopen it.
+	std::thread::sleep(Duration::from_millis(1000));
```

Comment:
Look like the magic sleep's necessary.

I tried a loop:

```rust
        let timestamp = std::time::Instant::now();
        loop {
                let settings = DatabaseSettings {
                        trie_cache_maximum_size,
                        state_pruning: Some(PruningMode::ArchiveAll),
                        source: DatabaseSource::ParityDb { path: path.clone() },
                        blocks_pruning: BlocksPruning::KeepAll,
                };

                match Backend::new(settings, 100) {
                        Ok(backend) => return backend,
                        Err(error) => {
                                if timestamp.elapsed() > Duration::from_secs(30) {
                                        // Timeout just in case, so we don't infinite loop.
                                        panic!(
                                                "failed to create a backend (waited for {}s): {}",
                                                timestamp.elapsed().as_secs_f64(),
                                                error
                                        );
                                } else {
                                        std::thread::sleep(Duration::from_millis(100));
                                        continue
                                }
                        },
                }
        }
```

But it doesn't seem to work:

```
thread 'main' panicked at 'failed to create a backend (waited for 30.038220199s): Backend error: Database file is in use. (Resource temporarily unavailable (os error 11))', client/db/benches/state_access.rs:198:21
```

So it looks like every call to `Backend::new` partially opens the database and doesn't close it fully before returning?

Reply:
Okay :(

Reply:
I might be able to get rid of it in another way though, not sure yet. I'm currently experimenting with slightly changing how these are run since the benchmarks are still not as reliable as I'd like (probably due to having to start the whole machinery around these state accesses, so that bleeds into the benchmarks and gets Criterion to output that we have slightly performance improvements/regressions even though nothing actually changed so there should have been no change).